aerogear/aerogear-sync-server
# SyncServer sends patch to all clients, even the client that initiated the patch.

Motivation:
Currently when a client sends a patch message to the server it should
not receive a patch message in response, unless the document changed on
the server after the client's patch has been applied.
At the moment the client is receiving a patch message regardless (though
the patch is empty as there is nothing to apply)

Jira:
https://issues.jboss.org/browse/AGSYNC-1

Steps to test (part from running the unit tests)
1. [Start](https://github.com/danbev/aerogear-sync-server/tree/refactoring-squashed/server/server-netty#jsonpatchsyncserver) the JsonPatchSyncServer. 
2. Check out @lholmquist [aerogear-js-cookbook](https://github.com/lholmquist/aerogear-js-cookbook/tree/diff-sync-json-patch) branch. Follow the instructions to start the ember server.

Notice in the logs that when using a single webpage with the Luke Skywalker demo open that when patches are sent to the server there, there is nothing sent to that client by the server. This can be seen by inspecting the console log and you should _not_ see a line like this (as long as there is only one client connected):
```shell
[nioEventLoopGroup-1-0] INFO org.jboss.aerogear.sync.server.ServerSyncEngine - Patched Document [12345] content: {"name":"Luke Skywalker","profession":"Jedi2","hobbies":[{"id":"0497aaec-002a-4a98-85e6-0928057b0d07","description":"Fighting the Dark Side"},{"id":"64ef749f-00ce-436c-9a8a-4c57bbfac8df","description":"going into Tosche Station to pick up some power converters"},{"id":"240950a8-aa2c-4596-b025-5292de2e6230","description":"Kissing his sister"},{"id":"d5460479-46b9-4ffc-ac6e-a229aed72976","description":"Bulls eyeing Womprats on his T-16"}]}
[nioEventLoopGroup-1-0] DEBUG org.jboss.aerogear.sync.server.ServerSyncEngine - Sending to [5ad4417e-5d7c-48d9-bb7a-18a3227cd6d1] : JsonPatchMessage[documentId=12345, clientId=5ad4417e-5d7c-48d9-bb7a-18a3227cd6d1, edits=[JsonPatctEdit[serverVersion=0, clientVersion=1, diff=JsonPatchDiff[jsonPatch=[]]]]]
```
The ```Sending to``` line is what we are looking for here. Since there have not been any other updates other than this clients updates, we are sending an empty patch. This does no harm and can act as an acknowledgment from the server. But this does cause extra network traffic which is unnecessary.
